kernel/smp.c

```diff
@@ -80,35 +98,48 @@ void z_smp_thread_swap(void)
 static inline FUNC_NORETURN void smp_init_top(void *arg)
 {
 	struct k_thread dummy_thread;
+	struct cpu_start_cb *csc = arg;
 
 	(void)atomic_set(&ready_flag, 1);
 
-	wait_for_start_signal(arg);
+	wait_for_start_signal(&cpu_start_flag);
 	z_dummy_thread_init(&dummy_thread);
 #ifdef CONFIG_SYS_CLOCK_EXISTS
 	smp_timer_init();
 #endif
 
+	/* Do additional initialization steps if needed. */
+	if ((csc != NULL) && (csc->fn != NULL)) {
+		csc->fn(csc->arg);
```

Comment:
@dcpleung isn't this racy in general? On all CPUs `csc` points to `cpu_start_fn`, right? And the CPU gets here after signalling the initiator CPU, the one that started this one. So the initiator CPU can go ahead and start yet another CPU potentially with different parameters. So far that shouldn't be a problem for SOF, because there `.fn` is the same for all and `.arg` is always `NULL`, but in general this seems wrong?

Reply:
Yes, there is potential for conflicts here. So the caller must take precautions to avoid that situation. Or we can implement a per-CPU data struct (but this is going to take up some space).